mesonbuild/compilers/cpp.py

```diff
@@ -0,0 +1,199 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .. import coredata
+
+from .compilers import *
+from . import CCompiler
+from . import VisualStudioCCompiler
+from . import GnuCompiler
+from . import ClangCompiler
+from . import IntelCompiler
+
+class CPPCompiler(CCompiler):
+    def __init__(self, exelist, version, is_cross, exe_wrap):
+        # If a child ObjCPP class has already set it, don't set it ourselves
+        if not hasattr(self, 'language'):
+            self.language = 'cpp'
+        CCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+
+    def get_no_stdinc_args(self):
+        return ['-nostdinc++']
+
+    def sanity_check(self, work_dir, environment):
+        code = 'class breakCCompiler;int main(int argc, char **argv) { return 0; }\n'
+        return self.sanity_check_impl(work_dir, environment, 'sanitycheckcpp.cc', code)
+
+    def get_compiler_check_args(self):
+        # -fpermissive allows non-conforming code to compile which is necessary
+        # for many C++ checks. Particularly, the has_header_symbol check is
+        # too strict without this and always fails.
+        return super().get_compiler_check_args() + ['-fpermissive']
+
+    def has_header_symbol(self, hname, symbol, prefix, env, extra_args=None, dependencies=None):
+        # Check if it's a C-like symbol
+        if super().has_header_symbol(hname, symbol, prefix, env, extra_args, dependencies):
+            return True
+        # Check if it's a class or a template
+        if extra_args is None:
+            extra_args = []
+        fargs = {'prefix': prefix, 'header': hname, 'symbol': symbol}
+        t = '''{prefix}
+        #include <{header}>
+        using {symbol};
+        int main () {{ return 0; }}'''
+        return self.compiles(t.format(**fargs), env, extra_args, dependencies)
+
+
+class VisualStudioCPPCompiler(VisualStudioCCompiler, CPPCompiler):
+    def __init__(self, exelist, version, is_cross, exe_wrap):
+        self.language = 'cpp'
+        CPPCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+        VisualStudioCCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+        self.base_options = ['b_pch'] # FIXME add lto, pgo and the like
+
+    def get_options(self):
+        return {'cpp_eh': coredata.UserComboOption('cpp_eh',
+                                                   'C++ exception handling type.',
+                                                   ['none', 'a', 's', 'sc'],
+                                                   'sc'),
+                'cpp_winlibs': coredata.UserStringArrayOption('cpp_winlibs',
+                                                              'Windows libs to link against.',
+                                                              msvc_winlibs)
```

Comment:
__[flake8]__
* [F405] 'msvc_winlibs' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->